feat: Add Rish-based installer with signature verification and downgrade support

juandvasquezp opened the pull request:
Hello, This is my first PR and I'm so excited about it.

# [What is rish](https://github.com/RikkaApps/Shizuku-API/blob/master/rish/README.md)
rish is an Android program for interacting with a shell that runs on a high-privileged daemon process. In a way, it allows to execute adb shell commands

A typical `rish` installation allows the user to use `rish` commands from everywhere by putting it inside `usr/bin/`

# What I did
I added `rish` support to Revancify in a way that allows installation of apps directly from the terminal, uninstall apps, get installed versions, in a similar way that the current `su` functionalities do.

# Details
- I added another flag `-R` to disable `rish`
- Inside `install.sh`, I modified so that when we have `rish` access, we'll use it.
- Inside `root.sh`, I added `rish` functionality to `getInstalledVersion`, so It'll work even without `su`, useful to check the current installation in the selection screen.
- Added a new `modules/system/rish.sh` that will handle the installation.
- Now `LAUNCH_APP_AFTER_MOUNT` works without su, it's the same implementation, but using `am force-stop` to stop the Keyboard and Termux instead of `kill -9`

# Installation Details
I verify the signature of the current installation (if there is), using `keytool`, so if there is a signature mismatch, the user can decide to continue (uninstalling the current app) or abort:
```
local STOCK_APP_PATH=$(rish -c 'pm path "'"$PKG_NAME"'" | sed -n "/base/s/package://p"')
local STOCK_APP_SIGNATURE=$(keytool -printcert -jarfile "$STOCK_APP_PATH" 2>/dev/null | awk '/SHA256:/{print $2}' | tr -d ':')
```

I check if the installation is a downgrade using the same command used in `import.sh`
```
    if jq -e '.[0] > .[1]' <<< "[\"${INSTALLED_VERSION:-0}\", \"$APP_VER\"]" &> /dev/null; then
```
The program checks if the user has set `ALLOW_APP_VERSION_DOWNGRADE` in Revancify configs, and decides to proceed or abort.

Choosing to continue with the downgrade, the program will try to install with two ways, using the `-d` flag from the [pm install](https://developer.android.com/tools/adb#pm) command (although it has never worked for me). If it fails, the script will try uninstalling the app and installing again.

If it isn't a downgrade or the stock app is installed, it'll install normally.

> Similar to the `su` version, the program uses mount_log.txt to log details.

# Testing
✅ Installed new app via rish
✅ Updated existing app with matching signature
✅ Prompted uninstall on signature mismatch
✅ Handled downgrade with config on/off
✅ Verified log creation and deletion of temp APK

# Video
https://github.com/user-attachments/assets/53bcb6ca-9bd7-4425-b8b7-665dda6ea60f

decipher3114 commented:
That is not an easy installation. User has to manually install python and python-pip to install this,
The copy paste method is easy and fast as it takes care of all the dependencies.

juandvasquezp committed:
feat: Add Rish-based installer with signature verification and downgrade support

Hello, This is my first PR and I'm so excited about it.

# [What is rish](https://github.com/RikkaApps/Shizuku-API/blob/master/rish/README.md)
rish is an Android program for interacting with a shell that runs on a high-privileged daemon process. In a way, it allows to execute adb shell commands

A typical `rish` installation allows the user to use `rish` commands from everywhere by putting it inside `usr/bin/`

# What I did
I added `rish` support to Revancify in a way that allows installation of apps directly from the terminal, uninstall apps, get installed versions, in a similar way that the current `su` functionalities do.

# Details
- I added another flag `-R` to disable `rish`
- Inside `install.sh`, I modified so that when we have `rish` access, we'll use it.
- Inside `root.sh`, I added `rish` functionality to `getInstalledVersion`, so It'll work even without `su`, useful to check the current installation in the selection screen.
- Added a new `modules/system/rish.sh` that will handle the installation.
- Now `LAUNCH_APP_AFTER_MOUNT` works without su, it's the same implementation, but using `am force-stop` to stop the Keyboard and Termux instead of `kill -9`

# Installation Details
I verify the signature of the current installation (if there is), using `keytool`, so if there is a signature mismatch, the user can decide to continue (uninstalling the current app) or abort:
```
local STOCK_APP_PATH=$(rish -c 'pm path "'"$PKG_NAME"'" | sed -n "/base/s/package://p"')
local STOCK_APP_SIGNATURE=$(keytool -printcert -jarfile "$STOCK_APP_PATH" 2>/dev/null | awk '/SHA256:/{print $2}' | tr -d ':')
```

I check if the installation is a downgrade using the same command used in `import.sh`
```
    if jq -e '.[0] > .[1]' <<< "[\"${INSTALLED_VERSION:-0}\", \"$APP_VER\"]" &> /dev/null; then
```
The program checks if the user has set `ALLOW_APP_VERSION_DOWNGRADE` in Revancify configs, and decides to proceed or abort.

Choosing to continue with the downgrade, the program will try to install with two ways, using the `-d` flag from the [pm install](https://developer.android.com/tools/adb#pm) command (although it has never worked for me). If it fails, the script will try uninstalling the app and installing again.

If it isn't a downgrade or the stock app is installed, it'll install normally.

> Similar to the `su` version, the program uses mount_log.txt to log details.

# Testing
✅ Installed new app via rish
✅ Updated existing app with matching signature
✅ Prompted uninstall on signature mismatch
✅ Handled downgrade with config on/off
✅ Verified log creation and deletion of temp APK

# Video
https://github.com/user-attachments/assets/53bcb6ca-9bd7-4425-b8b7-665dda6ea60f

decipher3114 commented:
This feature is amazing and very well implemented.

I have only one question.

Will it handle it properly if shizuku isn't running?

juandvasquezp commented:
> This feature is amazing and very well implemented.
> 
> I have only one question.
> 
> Will it handle it properly if shizuku isn't running?

I used the same command that checks if su is available, inside `revancify`, but for rish, to check if shizuku is available:

`rish -c 'exit'`

As you can see below
![Screenshot_20250528_230103_Termux.jpg](https://github.com/user-attachments/assets/43a347e2-96b3-4a90-9938-d9b78edf1104)

decipher3114 commented:
Please make the following changes:
- rename [./root](https://github.com/decipher3114/Revancify/tree/main/root) to `./system` or something more generic.
- move all the rish-install related code to `./system/rish-install.sh`
- make `rish-install.sh` consistent to the `mount.sh` file and not the modules.
- make `./modules/system/rish.sh` for the functions used in querying.
- don't uninstall app to downgrade (this leads to loss of user data), instead just show a warning and exit.

juandvasquezp committed:
feat: Add Rish-based installer with signature verification and downgrade support  (#2)

# New changes
I reorganized the code to be even more robust, with a similar structure than `su`:
- User queries and logic is in `.modules/system/rish.sh`
- Renamed `.root/` to `.system/` and changed the paths of the new scripts inside `.modules/system/root.sh` accordingly
- The install and uninstall scripts are in `.system/rish-install.sh` and `.system/rish-uninstall.sh`, they are only there to execute the functions and verify.
- Added more queries to the user.

# Tested workflows
✅ Installed new app
✅ Updated existing app with matching signature
✅ Prompted uninstall on signature mismatch, ask permission to te user to uninstall current installation
- ✅ User says yes
- ✅ User says no, abort
✅ Handled downgrade with config on/off, if downgrades are allowed,  ask permission to te user to uninstall current installation
- ✅ User says yes
- ✅ User says no, abort
✅ Verified log creation and deletion of temp APK

# Multiple users installations
When I was testing with some apps, I found some weird errors related to more users, it was because I had the same package name installed in my secure folder. So if one of the workflows above fails (downgrade or signature mismatch) , I checked for hidden apps with `dumpsys package <package>`, and prompted the user for permission to uninstall the app from all users.
![Captura desde 2025-05-30 08-17-26](https://github.com/user-attachments/assets/a6c5190a-f565-4833-9971-7f567ac2c0c1)

### Why not uninstall from all users directly?
Some apps installed directly from my secured folder couldn't be uninstalled with `rish` , and required manual uninstallation from the user. Also some times I could have a stock app in my secure folder and a patched app in my normal user, but most of the time I couldn't have mismatched apps in each user.
So I found it very inconsistent, so I did that if we encounter a hidden app, we do a second attempt if the user allows it, if we can leave the other users intact, the better.

# Video
In this video I have in both my user and in the secure folder the stock PlayStore version or RAR:

https://github.com/user-attachments/assets/5ed6c3eb-f517-4d8f-9258-6c724cb2c6ea

juandvasquezp commented:
# New changes
I reorganized the code to be even more robust, with a similar structure than `su`:
- User queries and logic is in `.modules/system/rish.sh`
- Renamed `.root/` to `.system/` and changed the paths of the new scripts inside `.modules/system/root.sh` accordingly
- The install and uninstall scripts are in `.system/rish-install.sh` and `.system/rish-uninstall.sh`, they are only there to execute the functions and verify.
- Added more queries to the user.

# Tested workflows
✅ Installed new app
✅ Updated existing app with matching signature
✅ Prompted uninstall on signature mismatch, ask permission to te user to uninstall current installation
- ✅ User says yes
- ✅ User says no, abort
✅ Handled downgrade with config on/off, if downgrades are allowed,  ask permission to te user to uninstall current installation
- ✅ User says yes
- ✅ User says no, abort
✅ Verified log creation and deletion of temp APK

# Multiple users installations
When I was testing with some apps, I found some weird errors related to more users, it was because I had the same package name installed in my secure folder. So if one of the workflows above fails (downgrade or signature mismatch) , I checked for hidden apps with `dumpsys package <package>`, and prompted the user for permission to uninstall the app from all users.
![Captura desde 2025-05-30 08-17-26](https://github.com/user-attachments/assets/a6c5190a-f565-4833-9971-7f567ac2c0c1)

### Why not uninstall from all users directly?
Some apps installed directly from my secured folder couldn't be uninstalled with `rish` , and required manual uninstallation from the user. Also some times I could have a stock app in my secure folder and a patched app in my normal user, but most of the time I couldn't have mismatched apps in each user.
So I found it very inconsistent, so I did that if we encounter a hidden app, we do a second attempt if the user allows it, if we can leave the other users intact, the better.

# Video
In this video I have in both my user and in the secure folder the stock PlayStore version or RAR:

https://github.com/user-attachments/assets/5ed6c3eb-f517-4d8f-9258-6c724cb2c6ea

juandvasquezp committed:
A little cleanup, idk why I used single quotes instead of double ones (#3)

* A little cleanup, I really don't know why I set myself to use single quotes instead of double

juandvasquezp committed:
Typo in the automatic launch, setting it the same as su version (#4)
